### PR TITLE
WT-9474 Implement cursor bound checking for cursor next / prev for column store.

### PR DESCRIPTION
**Pull request by jiechenbo**

### CODE REVIEW

**Comment by luke-pearson** [comment]
Once #8082 is merged we'll update this PR then review.

File: src/btree/bt_curnext.c [hunk 1]
```diff
@@ -172,6 +174,16 @@ __cursor_var_append_next(WT_CURSOR_BTREE *cbt, bool newpage, bool restart, size_
             return (WT_NOTFOUND);
 
         __cursor_set_recno(cbt, WT_INSERT_RECNO(cbt->ins));
+
+        /*
+         * If an upper bound has been set ensure that the key is within the range, otherwise early
+         * exit.
+         */
+        if ((ret = __wt_btcur_bounds_early_exit(session, cbt, true, key_out_of_boundsp)) ==
```
Comment (luke-pearson) [inline src/btree/bt_curnext.c:182]: I believe we want this after the `restart_read` section, thus if we return a prepare conflict we don't accidentally skip checking the bound after restarting the read.

File: src/btree/bt_curnext.c [hunk 2]
```diff
@@ -239,6 +253,15 @@ __cursor_var_next(WT_CURSOR_BTREE *cbt, bool newpage, bool restart, size_t *skip
         __cursor_set_recno(cbt, cbt->recno + 1);
 
 new_page:
+        /*
```
Comment (luke-pearson) [inline src/btree/bt_curnext.c:256]: Similarly here.

File: src/btree/bt_curnext.c [hunk 3]
```diff
@@ -239,6 +253,15 @@ __cursor_var_next(WT_CURSOR_BTREE *cbt, bool newpage, bool restart, size_t *skip
...
+         * If an upper bound has been set ensure that the key is within the range, otherwise early
+         * exit.
+         */
+        if ((ret = __wt_btcur_bounds_early_exit(session, cbt, true, key_out_of_boundsp)) ==
+          WT_NOTFOUND)
+            WT_STAT_CONN_DATA_INCR(session, cursor_bounds_next_early_exit);
+        WT_RET(ret);
+
```
Comment (luke-pearson) [inline src/btree/bt_curnext.c:264]: Given that the recno is modified later in this function do we need to check again?

File: src/btree/bt_curnext.c [hunk 4]
```diff
@@ -821,13 +844,23 @@ __wt_btcur_next_prefix(WT_CURSOR_BTREE *cbt, WT_ITEM *prefix, bool truncating)
             F_CLR(cbt, WT_CBT_ITERATE_APPEND);
             if (ret != WT_NOTFOUND)
                 break;
+
+            /*
+             * If we are doing a search near with prefix key configured or the cursor has bounds
```
Comment (luke-pearson) [inline src/btree/bt_curnext.c:849]: Please remove the prefix key stuff from this comment.

File: src/btree/bt_cursor.c [hunk 5]
```diff
@@ -65,8 +72,6 @@ __btcur_bounds_search_near_reposition(WT_SESSION_IMPL *session, WT_CURSOR_BTREE
     key_out_of_bounds = upper = false;
 
     WT_ASSERT(session, WT_CURSOR_BOUNDS_SET(cursor));
-    if (CUR2BT(cbt)->type != BTREE_ROW)
```
Comment (luke-pearson) [inline src/btree/bt_cursor.c:68]: Given that we removed this check entirely does the code work for FLCS?

**Comment by luke-pearson** [comment]
All my comments in cursor->next also apply for prev.

File: src/btree/bt_cursor.c (hunk 5 above)
Comment (jiechenbo) [reply src/btree/bt_cursor.c:68]: No, I have added a check we can't set bounds on a btree if it is FLCS

File: src/btree/bt_cursor.c [hunk 6]
```diff
@@ -2223,13 +2236,13 @@ __wt_btcur_close(WT_CURSOR_BTREE *cbt, bool lowlevel)
 }
 
 /*
- * __wt_btcur_bounds_row_position --
+ * __wt_btcur_bounds_position --
  *     A unpositioned bounded cursor need to start its cursor next and prev walk from the lower or
```
Comment (clarissecheah) [inline src/btree/bt_cursor.c:2240]: np: Not a part of this PR, but could you fix up the comment here? "An unpositioned bounded cursor needs to start its cursor next and prev walk..."

File: src/include/btree_cmp_inline.h [hunk 7]
```diff
@@ -138,31 +138,55 @@ __wt_prefix_match(const WT_ITEM *prefix, const WT_ITEM *tree_item)
 }
 
 /*
- * __wt_row_compare_bounds --
- *     Return if the cursor key is within the bounded range. If next is True, this indicates a next
- *     call and the key is checked against the upper bound. If next is False, this indicates a prev
- *     call and the key is then checked against the lower bound.
+ * __wt_compare_bounds --
+ *     Return if the cursor key is within the bounded range. If upper is True, this indicates a
+ *     upper call and the key is checked against the upper bound. If upper is False, this indicates
```
Comment (clarissecheah) [inline src/include/btree_cmp_inline.h:143]: np: "an upper call"

File: src/btree/bt_curnext.c [hunk 8]
```diff
@@ -172,7 +174,17 @@ __cursor_var_append_next(WT_CURSOR_BTREE *cbt, bool newpage, bool restart, size_
             return (WT_NOTFOUND);
 
         __cursor_set_recno(cbt, WT_INSERT_RECNO(cbt->ins));
+
 restart_read:
+        /*
```
Comment (luke-pearson) [inline src/btree/bt_curnext.c:179]: We're going to need testing around this restart_read edgecase.

File: src/btree/bt_curprev.c [hunk 9]
```diff
@@ -769,6 +791,7 @@ __wt_btcur_prev(WT_CURSOR_BTREE *cbt, bool truncating)
             F_CLR(cbt, WT_CBT_ITERATE_APPEND);
             if (ret != WT_NOTFOUND)
                 break;
+
```
Comment (luke-pearson) [inline src/btree/bt_curprev.c:794]: Remove this whitespace.

File: src/btree/bt_curnext.c [hunk 10]
```diff
@@ -172,7 +174,17 @@ __cursor_var_append_next(WT_CURSOR_BTREE *cbt, bool newpage, bool restart, size_
             return (WT_NOTFOUND);
 
         __cursor_set_recno(cbt, WT_INSERT_RECNO(cbt->ins));
+
```
Comment (luke-pearson) [inline src/btree/bt_curnext.c:177]: Remove this whitespace.

File: src/btree/bt_curnext.c (hunk 8 above)
Comment (luke-pearson) [reply src/btree/bt_curnext.c:179]: Can be done in another ticket if required.

File: src/btree/bt_curprev.c [hunk 11]
```diff
@@ -789,16 +812,18 @@ __wt_btcur_prev(WT_CURSOR_BTREE *cbt, bool truncating)
...
-             * mismatch or the key is out of bounds. If so, we break instead of walking onto the
-             * next page. We're not directly returning here to allow the cursor to be reset first
-             * before we return WT_NOTFOUND.
-             */
-            if (key_out_of_bounds)
-                break;
         }
+
+        /*
+         * If we are doing a search near with prefix key configured or the cursor has bounds set, we
```
Comment (luke-pearson) [inline src/btree/bt_curprev.c:818]: bt_curprev has no notion of prefix search so please remove it from this comment.

File: src/btree/bt_curnext.c (hunk 8 above)
Comment (jiechenbo) [reply src/btree/bt_curnext.c:179]: This will be done in WT-9475, with writing the prepared transactions test

**Review by luke-pearson — APPROVED** [review]
LGTM, will fix fuzz test before we merge.

**Review by clarissecheah — APPROVED** [review]
LGTM.

**Comment by jiechenbo** [comment]
evergreen merge

**Comment by jiechenbo** [comment]
evergreen merge

**Pull request merged**